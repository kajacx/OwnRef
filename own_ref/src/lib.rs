mod macros;
mod structs;

pub use structs::OwnRef;
pub use structs::OwnRefMut;
