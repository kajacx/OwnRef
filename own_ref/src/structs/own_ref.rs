use std::marker::PhantomData;
use std::ops::Deref;
use std::ptr::NonNull;

pub struct OwnRef<'a, T> {
    pointer: NonNull<T>,
    _phantom_lifetime: PhantomData<&'a ()>,
    _phantom_value: PhantomData<T>,
}

impl<'a, T> OwnRef<'a, T> {
    pub(crate) unsafe fn new<'b>(reference: &'b T, _phantom_lifetime: PhantomData<&'a ()>) -> Self {
        Self {
            pointer: NonNull::from(reference),
            _phantom_lifetime,
            _phantom_value: PhantomData,
        }
    }

    pub(crate) fn from_pointer(
        pointer: NonNull<T>,
        _phantom_lifetime: PhantomData<&'a ()>,
    ) -> Self {
        Self {
            pointer,
            _phantom_lifetime,
            _phantom_value: PhantomData,
        }
    }

    pub fn get(&self) -> &T {
        unsafe { self.pointer.as_ref() }
    }

    pub fn take(self) -> T {
        let result = unsafe { std::ptr::read(self.pointer.as_ptr()) };
        std::mem::forget(self); // do not run destructor
        result
    }
}

impl<'a, T> Deref for OwnRef<'a, T> {
    type Target = T;

    fn deref(&self) -> &Self::Target {
        self.get()
    }
}

impl<T> Drop for OwnRef<'_, T> {
    fn drop(&mut self) {
        let _drop = unsafe { std::ptr::read(self.pointer.as_ptr()) };
    }
}
