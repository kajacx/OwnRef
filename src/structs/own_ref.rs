use std::marker::PhantomData;
use std::ops::Deref;
use std::ptr::NonNull;

pub struct OwnRef<'a, T> {
    pointer: NonNull<T>,
    _phantom_lifetime: PhantomData<&'a ()>,
    _phantom_value: PhantomData<T>,
}

impl<'a, T> OwnRef<'a, T> {
    pub unsafe fn new<'b>(reference: &'b T, _phantom_lifetime: &'a ()) -> Self {
        Self {
            pointer: NonNull::from(reference),
            _phantom_lifetime: PhantomData,
            _phantom_value: PhantomData,
        }
    }

    pub fn get(&self) -> &T {
        unsafe { self.pointer.as_ref() }
    }

    pub fn take(self) -> T {
        let result = unsafe { std::ptr::read(self.pointer.as_ptr()) };
        std::mem::forget(self); // do not run destructor
        result
    }
}

impl<'a, T> Deref for OwnRef<'a, T> {
    type Target = T;

    fn deref(&self) -> &Self::Target {
        self.get()
    }
}

impl<T> Drop for OwnRef<'_, T> {
    fn drop(&mut self) {
        let _drop = unsafe { std::ptr::read(self.pointer.as_ptr()) };
    }
}
