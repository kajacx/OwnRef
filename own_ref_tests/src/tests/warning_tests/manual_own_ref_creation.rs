use own_ref::{lifetime_of, OwnRef};

fn main() {
    let text = "Hello".to_string();

    let _phantom = ();
    let owned_ref = unsafe { OwnRef::new(&text, lifetime_of(&_phantom)) };
    std::mem::forget(text);

    assert_eq!(owned_ref.as_str(), "Hello");
}

fn fail() -> i32 {
    // TODO: to check the message of the warning, the compilation needs to fail.

    // Only then will trybuild check the (errors and) warnings.
}
