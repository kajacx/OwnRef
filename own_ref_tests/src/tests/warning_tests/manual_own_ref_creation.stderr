warning: use of deprecated associated function `own_ref::OwnRef::<'a, T>::new`: Do not construct OwnRef manually, instead use the `new_own_ref` macro.
 --> src/tests/warning_tests/manual_own_ref_creation.rs:7:38
  |
7 |     let owned_ref = unsafe { OwnRef::new(&text, lifetime_of(&_phantom)) };
  |                                      ^^^
  |
  = note: `#[warn(deprecated)]` on by default

error[E0308]: mismatched types
  --> src/tests/warning_tests/manual_own_ref_creation.rs:13:14
   |
13 | fn fail() -> i32 {
   |    ----      ^^^ expected `i32`, found `()`
   |    |
   |    implicitly returns `()` as its body has no tail or `return` expression
